validator/errors.py

```diff
@@ -0,0 +1,125 @@
+""" Error-code and schema related constants. """
+
+from typing import Mapping
+
+from cerberus.errors import (BasicErrorHandler, ErrorDefinition, ErrorTree,
+                             ValidationError)
+
+
+# pylint: disable=(too-few-public-methods)
+class ErrorDefs:
+    """ Class to define custom errors. """
+    # IMPORTANT - When adding a new error DON'T change the existing codes
+    # Error codes are used to map b/w cerberus erros and NACC error codes in QC gear
+    # Cerberus uses bit 5 and bit 7 to mark specific error types
+    # Check https://docs.python-cerberus.org/customize.html for more info
+    CURR_DATE_MAX = ErrorDefinition(0x1000, 'max')
+    CURR_YEAR_MAX = ErrorDefinition(0x1001, 'max')
+    INVALID_DATE_MAX = ErrorDefinition(0x1002, 'max')
+    CURR_DATE_MIN = ErrorDefinition(0x1003, 'min')
+    CURR_YEAR_MIN = ErrorDefinition(0x1004, 'min')
+    INVALID_DATE_MIN = ErrorDefinition(0x1005, 'max')
+    FILLED_TRUE = ErrorDefinition(0x1006, 'filled')
+    FILLED_FALSE = ErrorDefinition(0x1007, 'filled')
+    COMPATIBILITY_TRUE = ErrorDefinition(0x1008, 'compatibility')
+    COMPATIBILITY_FALSE = ErrorDefinition(0x1009, 'compatibility')
+    TEMPORAL = ErrorDefinition(0x2000, 'temporalrules')
+    NO_PRIMARY_KEY = ErrorDefinition(0x2001, 'temporalrules')
+    NO_PREV_VISIT = ErrorDefinition(0x2002, 'temporalrules')
+    FORMULA = ErrorDefinition(0x2003, 'logic')
+    CHECK_GDS_1 = ErrorDefinition(0x2004, 'check_with')
+    CHECK_GDS_2 = ErrorDefinition(0x2005, 'check_with')
+    CHECK_GDS_3 = ErrorDefinition(0x2006, 'check_with')
+    CHECK_GDS_4 = ErrorDefinition(0x2007, 'check_with')
+    CHECK_GDS_5 = ErrorDefinition(0x2008, 'check_with')
+
+
+class CustomErrorHandler(BasicErrorHandler):
+    """ Class to provide custom error messages  """
+
+    def __init__(self, schema: Mapping = None, tree: ErrorTree = None):
+        """
+
+        Args:
+            schema (optional): Validation schema as dict[field, rule objects].
+            tree (optional): Validation errors tree.
+        """
+
+        super().__init__()
+        self.__set_custom_error_codes()
+        self._custom_schema = schema
+
+    def __set_custom_error_codes(self):
+        """Add custom error codes specific to NACC validation rules
+        """
+        # IMPORTANT - When adding a new error DON'T change the existing codes
+        # Cerberus uses bit 5 and bit 7 to mark specific error types
+        # Check https://docs.python-cerberus.org/customize.html for more info
+        # Error messages are synced with ErrorDefs using the error code
+        # Error codes are also used to map b/w cerberus erros and NACC error codes in QC gear
+        custom_errors = {
```

Comment:
Is this dict going to grow as new rules are added? And, if so is there a way to handle them dynamically

Reply:
Any new rule definition using the existing rule types can be added without modifying the code. If a new rule type needs to be added, validator code has to be updated to support the new rule type, and the errors dict will also be extended alongside.